pxr/imaging/hdSt/material.cpp

```diff
@@ -172,48 +177,91 @@ HdStMaterial::_ProcessTextureDescriptors(
 /* virtual */
 void
 HdStMaterial::Sync(HdSceneDelegate *sceneDelegate,
-                 HdRenderParam   *renderParam,
-                 HdDirtyBits     *dirtyBits)
+                   HdRenderParam   *renderParam,
+                   HdDirtyBits     *dirtyBits)
 {
     HD_TRACE_FUNCTION();
     HF_MALLOC_TAG_FUNCTION();
 
+    const HdDirtyBits bits = *dirtyBits;
+
+    if (!(bits & DirtyResource) && !(bits & DirtyParams)) {
+        *dirtyBits = Clean;
+        return;
+    }
+
+    bool processedMaterialNetwork = false;
+
     HdStResourceRegistrySharedPtr const& resourceRegistry =
         std::static_pointer_cast<HdStResourceRegistry>(
             sceneDelegate->GetRenderIndex().GetResourceRegistry());
 
-    HdDirtyBits bits = *dirtyBits;
+#ifdef PXR_MATERIALX_SUPPORT_ENABLED
+    {
+        HdStRenderDelegate* stormDelegate = static_cast<HdStRenderDelegate*>(
+            sceneDelegate->GetRenderIndex().GetRenderDelegate());
 
-    if (!(bits & DirtyResource) && !(bits & DirtyParams)) {
-        *dirtyBits = Clean;
-        return;
+        HdSt_MaterialFilterTaskSharedPtr filterTask;
+
+        // This scene index manages early parallel MaterialX codegen. It exists
+        // when that optimization is enabled.
+        if (HdSt_MaterialXSyncSceneIndex* sceneIndex =
+            stormDelegate->GetMaterialXSyncSceneIndex()) {
+
+            // Wait for all early parallel codegen tasks to complete and
+            // retrieve the state cached for this sprim when codegen was
+            // started
+            filterTask = sceneIndex->WaitAndExtractFilterTask(GetId());
+        }
+
+        if (filterTask) {
+            // We only use filter tasks for non-volume materials
+            constexpr bool isVolume = false;
+
+            _hdStMaterialNetwork.ProcessFilterTask(
+                GetId(), filterTask,
+                isVolume, resourceRegistry.get());
+
+            processedMaterialNetwork = true;
+        }
     }
+#endif
 
-    bool markBatchesDirty = false;
+    // The serial MaterialX codegen code path - executed only if we haven't
+    // already processed the material network above
+    if (!processedMaterialNetwork) {
+        VtValue vtMat = sceneDelegate->GetMaterialResource(GetId());
+        if (vtMat.IsHolding<HdMaterialNetworkMap>()) {
+
+            HdMaterialNetworkMap const& hdNetworkMap =
+                vtMat.UncheckedGet<HdMaterialNetworkMap>();
+
+            if (!hdNetworkMap.terminals.empty() && !hdNetworkMap.map.empty()) {
+                _hdStMaterialNetwork.ProcessMaterialNetwork(GetId(),
+                    hdNetworkMap, resourceRegistry.get());
+
+                processedMaterialNetwork = true;
```

Comment:
Once the material network has processed the filter task, we don't need the task any more, so we destroy it by letting it go out of scope.